File: src/main/java/features/reports/creators/RepoAccessorReportCreator.java
package features.reports.creators;

import core.algorithm.Algorithm;
import core.data.DataTable;
import core.data.medimage.ImageSeries;
import core.data.medimage.MirfAttributes;
import core.data.report.AlgorithmReport;
import core.data.report.DataTableAlgorithmReport;
import core.repository.RepositoryInfo;
import features.repositoryaccessors.RepoAccessorsAttributes;
import features.repositoryaccessors.RepositoryRequestInfo;

import java.util.ArrayList;
import java.util.Dictionary;
import java.util.Hashtable;
import java.util.List;

public class RepoAccessorReportCreator implements Algorithm<ImageSeries, AlgorithmReport> {

    //TODO: (avlomakin) replace constants with resource variables
    private static final String REPOSITORY_HEADER = "Repository";
    private static final String LINK_HEADER = "Link";
    private static final String TOTAL_LOADED = "Total loaded";
    private static final String IMAGE_TYPE_HEADER = "Image type";


    private List<Dictionary<String, String>> getRows(ImageSeries medImages) {

        RepositoryInfo repositoryInfo = medImages.findAttributeValue(MirfAttributes.REPO_INFO.tag);
        RepositoryRequestInfo requestInfo = medImages.findAttributeValue(RepoAccessorsAttributes.REPOSITORY_REQUEST_INFO.tag);

        String totalLoaded = String.valueOf(medImages.images.size());

        Dictionary<String, String> row = new Hashtable<>();
        row.put(REPOSITORY_HEADER, repositoryInfo.repositoryName);
        row.put(LINK_HEADER, requestInfo.link);
        row.put(TOTAL_LOADED, totalLoaded);
        row.put(IMAGE_TYPE_HEADER, medImages.images.get(0).getExtension());

        return new ArrayList<Dictionary<String, String>>(){{add(row);}};
    }

    private ArrayList<String> getHeaders() {
        return new ArrayList<String>(){{
            add(REPOSITORY_HEADER);
            add(LINK_HEADER);
            add(TOTAL_LOADED);
            add(IMAGE_TYPE_HEADER);
        }};
    }

    @Override
    public AlgorithmReport execute(ImageSeries medImages) {
        DataTable reportTable = new DataTable();
        reportTable.columns.addAll(getHeaders());
        reportTable.rows.addAll(getRows(medImages));

        DataTableAlgorithmReport report = new DataTableAlgorithmReport(reportTable);
        return report;
    }
}

File: src/main/kotlin/com/mirf/features/reports/creators/RepoAccessorReportCreator.kt
package com.mirf.features.reports.creators

import com.mirf.core.algorithm.Algorithm
import com.mirf.core.data.DataTable
import com.mirf.core.data.medimage.ImageSeries
import com.mirf.core.data.attribute.MirfAttributes
import com.mirf.core.data.report.AlgorithmReport
import com.mirf.core.data.report.DataTableAlgorithmReport
import com.mirf.features.repositoryaccessors.RepoAccessorsAttributes

import java.util.ArrayList
import java.util.Dictionary
import java.util.Hashtable

/**
 * [AlgorithmReport] creator for RepositoryAccessors
 */
class RepoAccessorReportCreator : Algorithm<ImageSeries, AlgorithmReport> {

    private val headers: ArrayList<String>
        get() = object : ArrayList<String>() {
            init {
                add(REPOSITORY_HEADER)
                add(LINK_HEADER)
                add(TOTAL_LOADED)
                add(IMAGE_TYPE_HEADER)
            }
        }

    private fun getRows(medImages: ImageSeries): List<HashMap<String, String>> {

        val repositoryInfo = medImages.attributes.findAttributeValue(MirfAttributes.REPO_INFO)

        val requestInfo = medImages.attributes.findAttributeValue(RepoAccessorsAttributes.REPOSITORY_REQUEST_INFO)

        val totalLoaded = medImages.images.size.toString()

        val row = hashMapOf<String, String>()
        row[REPOSITORY_HEADER] = repositoryInfo!!.repositoryName
        row[LINK_HEADER] = requestInfo!!.link
        row[TOTAL_LOADED] = totalLoaded
        row[IMAGE_TYPE_HEADER] = medImages.images[0].extension

        return object : ArrayList<HashMap<String, String>>() {
            init {
                add(row)
            }
        }
    }

    override fun execute(input: ImageSeries): DataTableAlgorithmReport {
        val reportTable = DataTable()
        reportTable.columns.addAll(headers)
        reportTable.rows.addAll(getRows(input))

        return DataTableAlgorithmReport(reportTable)
    }

    companion object {

        //TODO: (avlomakin) replace constants with resource variables
        private const val REPOSITORY_HEADER = "Repository"
        private const val LINK_HEADER = "Link"
        private const val TOTAL_LOADED = "Total loaded"
        private const val IMAGE_TYPE_HEADER = "Image type"
    }
}
